src/sources/exec.rs

```diff
@@ -0,0 +1,830 @@
+    pid: Option<u32>,
+    exit_status: Option<i32>,
+    exec_duration_millis: &Option<u128>,
+) -> Event {
+    emit!(ExecEventReceived {
+        command: config.command_line().as_str(),
+        byte_size: line.len(),
+    });
+    let mut log_event = LogEvent::default();
+
+    // Add message
+    log_event.insert(log_schema().message_key(), line);
+
+    // Add timestamp
+    log_event.insert(log_schema().timestamp_key(), Utc::now());
+
+    // Add source type
+    log_event.insert(log_schema().source_type_key(), Bytes::from(EXEC));
+
+    // Add data stream of stdin or stderr (if needed)
+    if let Some(data_stream) = data_stream {
+        log_event.insert(DATA_STREAM_KEY, data_stream.clone());
+    }
+
+    // Add pid (if needed)
+    if let Some(pid) = pid {
+        log_event.insert(PID_KEY, pid as i64);
+    }
+
+    // Add exit status (if needed)
+    if let Some(exit_status) = exit_status {
+        log_event.insert(EXIT_STATUS_KEY, exit_status as i64);
+    }
+
+    // Add exec duration millis (if needed)
+    if let Some(exec_duration_millis) = exec_duration_millis {
+        log_event.insert(EXEC_DURATION_MILLIS_KEY, *exec_duration_millis as i64);
+    }
+
+    // Add hostname (if needed)
+    if let Some(hostname) = hostname {
+        log_event.insert(log_schema().host_key(), hostname.clone());
+    }
+
+    // Add command
+    log_event.insert(COMMAND_KEY, config.command.clone());
+
+    Event::Log(log_event)
+}
+
+fn spawn_reader_thread<R: 'static + AsyncRead + Unpin + std::marker::Send>(
+    reader: BufReader<R>,
+    shutdown: ShutdownSignal,
+    event_per_line: bool,
+    buf_size: u64,
+    stream: &'static str,
+    sender: Sender<(Bytes, &'static str)>,
+) {
+    // Start the green background thread for collecting
+    Box::pin(tokio::spawn(async move {
+        debug!("Start capturing {} command output.", stream);
+
+        let mut read_buffer: Vec<u8> = Vec::new();
+
+        let reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
+
+        let mut limited_reader = reader.take(buf_size);
+
+        if event_per_line {
+            // Keep reading lines (lines longer than the max buffer will be split)
+            while let Ok(bytes_read) = limited_reader.read_until(b'\n', &mut read_buffer).await {
+                if bytes_read == 0 {
+                    // If we get a continuous stream of \n the bytes_read will be at least 1
+                    debug!("End of input reached, stop reading.");
+                    break;
+                }
+
+                // Strip of the end of line bytes
+                if read_buffer.ends_with(&[b'\n']) {
+                    let _ = read_buffer.pop();
+
+                    if read_buffer.ends_with(&[b'\r']) {
+                        let _ = read_buffer.pop();
+                    }
+                }
+
+                let read_bytes = Bytes::from(read_buffer.clone());
+                if sender.send((read_bytes, stream)).await.is_err() {
```

Comment:
Under what conditions would this error? Should we be logging something here?

Reply:
I have added a log message. This will happen If the receive half of the channel is closed, either due to close being called or the Receiver handle dropping.